keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,258 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Scalability and performance of kube-apiserver is crucial for scalability
+of the whole Kubernetes cluster. Given that kube-apiserver is cpu-intensive
+process, scaling a single instance of it translates to optimizing amount
+of work is needed to process a request (cpu cycles and amount of allocated
+memory, as memory management is significant part of work done be
+kube-apiserver).
+
+This proposal is aiming to significantly reduce amount of work spent on
+serializing objects as well as amount of allocated memory to process that.
+
+## Motivation
+
+Running different types of scalability tests and analyzing large production
+clusters proves that large number of watches watching the same set of objects
+may cause significant load on kube-apiserver. An extreme example of it is
+[#75294][], where creation of a single large Endpoints object (almost 1MB of
+size, due to 5k pods backing it) in 5k-node cluster can completely overload
+kube-apiserver for 5 seconds.
+
+The main reason for that is that for every watcher (Endpoints are being watched
+by kube-proxy running on every one) kube-apiserver independently serializes
+(which also requires deep-copy) every single object being send via this watch.
+
+While this problem is extremely visible for watch, the situation looks the same
+for regular GET/LIST operations - reading the same object N times will result
+in serializing that N times independently.
+
+This proposal presents a solution for that problem.
+
+[#75294]: https://github.com/kubernetes/kubernetes/issues/75294
+
+### Goals
+
+- Reduce load on kube-apiserver, by not serializing the same object multiple
+times wherever it is possible.
+
+### Non-Goals
+
+- Change overall architecture of the system, by changing what data is being
+read/watched by different components.
+
+## Proposal
+
+This proposal does not introduce any user-visible changes - the proposed changes
+are purely implementational detail of kube-apiserver.
+
+We propose introducing an internal type implementing `runtime.Object` interface,
```

Comment:
Can you add discussion of alternatives? E.g., what about keeping an explicit LRU serialization cache where the key is type & resource version; and modifying the serialization path logic to optionally take such a cache.

Reply:
Actually there could be a cache per resource, so the key just needs to be RV.

Keeping the serialization cache and watch cache separate would allow one to enable the serialization cache without turning on the watch cache.

Reply:
> Can you add discussion of alternatives? E.g., what about keeping an explicit LRU serialization cache where the key is type & resource version; and modifying the serialization path logic to optionally take such a cache.

Heh.  I had the same question above :)

Reply:
See the bottom of the document ("Alternatives" section) :-)
I'm not convinced LRU cache is better, but potentially may be convinced...

I'm happy to have this discussion below in "Alternatives" section.

Reply:
Resolving given there is "Alternatives" section in the bottom of the doc.